### PR TITLE
Engine: improve diagnostics

### PR DESCRIPTION
**Pull request by W95Psp**
This PR:
 - fixes the JSON Schema script for enum serialization;
 - adds some tests for the script;
 - simplify & change the phases and subtype module a bit to propagate spans everywhere so that we can report errors nicely

### CODE REVIEW

File: engine/utils/ocaml_of_json_schema/test/rust/src/lib.rs
Comment (franziskuskiefer) [inline engine/utils/ocaml_of_json_schema/test/rust/src/lib.rs:1]: Please add some comments to all files in this crate what they are doing.

File: engine/utils/ocaml_of_json_schema/test/rust/src/tester.rs
Comment (franziskuskiefer) [inline engine/utils/ocaml_of_json_schema/test/rust/src/tester.rs:2]:
```suggestion
```

File: engine/utils/ocaml_of_json_schema/test/rust/Cargo.toml
Comment (franziskuskiefer) [inline engine/utils/ocaml_of_json_schema/test/rust/Cargo.toml:1]: Can you add a readme.md or something here to say something about what it is and how it should be used?

File: engine/utils/ocaml_of_json_schema/test/ocaml/bin/dune
Comment (franziskuskiefer) [inline engine/utils/ocaml_of_json_schema/test/ocaml/bin/dune:20]:
```suggestion
   (:standard -w -A -w +8))))

```

**Review by franziskuskiefer — APPROVED** [review]
Thanks, lgtm to me after adding a little more docs.

File: engine/utils/ocaml_of_json_schema/test/rust/src/lib.rs
Comment (W95Psp) [reply engine/utils/ocaml_of_json_schema/test/rust/src/lib.rs:1]: I added a README that describes the thing, I think that's enough, what do you think?

Comment (franziskuskiefer) [reply engine/utils/ocaml_of_json_schema/test/rust/src/lib.rs:1]: Generally all the code needs comments as well (readme is high level doc also for someone just using it, comments in the code is for anyone working on the code including future selves).
I'll complain about any undocumented code but won't block on it for now in favour of speed.

Comment (W95Psp) [reply engine/utils/ocaml_of_json_schema/test/rust/src/lib.rs:1]: Yes, I agree, I basically wrote no comment while rushing adding features.
Ok, nice, thanks!

**Pull request merged**
